lib/rubocop/cop/performance/sample.rb

```diff
@@ -12,78 +12,107 @@ module Performance
       #   [1, 2, 3].sample
       #   [1, 2, 3].sample(3)
-      #   [1, 2, 3].sample(random: Random.new(1))
+      #   [1, 2, 3].shuffle(random: Random.new)
+      #   [1, 2, 3].shuffle(random: Random.new).find(&:odd?)
       class Sample < Cop
-        MSG = 'Use `sample` instead of `shuffle%s`.'
-        RANGE_TYPES = [:irange, :erange]
-        VALID_ARRAY_SELECTORS = [:first, :last, :[], nil]
+        MSG = 'Use `%<right>s` instead of `%<wrong>s`.'
+        OFFENDING_METHODS = [:first, :last, :[]]
 
         def on_send(node)
-          _receiver, first_method, params, = *node
-          return unless first_method == :shuffle
-          _receiver, second_method, params, = *node.parent if params.nil?
-          return unless VALID_ARRAY_SELECTORS.include?(second_method)
-          return if second_method.nil? && params.nil?
-
-          add_offense(node, range_of_shuffle(node), message(node, params))
+          analyzer = ShuffleAnalyzer.new(node)
+          return unless analyzer.offensive?
+          add_offense(node, analyzer.range_of_shuffle, analyzer.message)
         end
 
         def autocorrect(node)
-          _receiver, _method, params, selector = *node
-          _receiver, _method, params, selector = *node.parent if params.nil?
+          ShuffleAnalyzer.new(node).autocorrect
+        end
+
+        private
+
+        class ShuffleAnalyzer
+          def initialize(node)
+            @node   = node
+            @parent = node.parent
+          end
 
-          return if params && RANGE_TYPES.include?(params.type)
+          def autocorrect
+            _, _, params, selector = *parent
 
-          range = if params && (params.hash_type? || params.lvar_type?)
-                    range_of_shuffle(node)
-                  else
-                    Parser::Source::Range.new(node.loc.expression.source_buffer,
+            range = Parser::Source::Range.new(node.loc.expression.source_buffer,
                                               node.loc.selector.begin_pos,
-                                              node.parent.loc.selector.end_pos)
-                  end
+                                              parent.loc.selector.end_pos)
 
-          lambda do |corrector|
-            corrector.replace(range, 'sample')
-            return if selector.nil?
-            corrector.insert_after(range, "(#{selector.loc.expression.source})")
+            lambda do |corrector|
+              corrector.replace(range, 'sample')
+              if selector
+                corrector.insert_after(range, "(#{selector.loc.expression.source})")
+              elsif params && params.erange_type?
+                vals = params.children
+                size = vals.last.to_a.first - vals.first.to_a.first
+                corrector.insert_after(range, "(#{size})")
+              elsif params && params.irange_type?
+                vals = params.children
+                size = vals.last.to_a.first - vals.first.to_a.first + 1
```

Comment:
I realized that the last comment didn't make full sense. In order to use `Range#size`, you would have to extract the variables and use them to build a range to call `size` on. This will work when the elements of the range are explicit integers, but it will be non deterministic if the elements are assigned to variables.

```ruby
(1..10).to_a[0..5]
```
We can extract the `0` and the `5`, similar to what you have done, and be able to determine the range.

```ruby
start = 0
end = 5
(1..10).to_a[start..end]
```
Regardless or implementation, we would not be able to determine the size here because we will not know what `start` and `end` are.

Theoretically, RuboCop could scan through earlier code to try and determine what the variables are assigned to, but nothing like that, that I know of, exists today.